### PR TITLE
fix: broken traces

### PR DESCRIPTION
**Pull request by simaoribeiro**
# Description

Fixes broken traces by changing the Activity creation to `KafkaFlow` project, just before the messages are consumed and produced.

The error appeared because the Activity context, being created in `KafkaFlow.OpenTelemetry`, could not be propagated through the system and the next time a span would be created the existing Activity would be null, forcing the creation of a new trace instead of using the previous one

Closes https://github.com/Farfetch/kafkaflow/issues/468

## How Has This Been Tested?

Integration tests

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [x] I have added tests to cover my changes
-   [ ] I have made corresponding changes to the documentation

### Disclaimer

By sending us your contributions, you are agreeing that your contribution is made subject to the terms of our [Contributor Ownership Statement](https://github.com/Farfetch/.github/blob/master/COS.md)

### CODE REVIEW

**Review by ruiqbarbosa — APPROVED** [review]

File: src/KafkaFlow/Event.cs [hunk 1]
```diff
@@ -25,23 +25,27 @@ public IEventSubscription Subscribe(Func<TArg, Task> handler)
...
                         continue;
                     }
 
-                    await handler.Invoke(arg);
-                }
-                catch (Exception e)
-                {
-                    this.logHandler.Error("Error firing event", e, new { Event = this.GetType().Name });
-                }
+                    yield return handler.Invoke(arg).ContinueWith(t =>
```
Comment (joelfoliveira) [inline src/KafkaFlow/Event.cs:42]: Please fix the indentation on this block (after the foreach the indentation has one extra level)

File: src/KafkaFlow.IntegrationTests/Core/Middlewares/GzipMiddleware.cs [hunk 2]
```diff
@@ -1,13 +1,17 @@
...
+    using System.Diagnostics;
     using System.Threading.Tasks;
     using KafkaFlow.IntegrationTests.Core.Handlers;
 
     internal class GzipMiddleware : IMessageMiddleware
     {
         public async Task Invoke(IMessageContext context, MiddlewareDelegate next)
         {
-            MessageStorage.Add((byte[]) context.Message.Value);
+            var source = new ActivitySource("KafkaFlow.OpenTelemetry");
```
Comment (joelfoliveira) [inline src/KafkaFlow.IntegrationTests/Core/Middlewares/GzipMiddleware.cs:11]: Please rebase with master since this has changed in the other PR https://github.com/Farfetch/kafkaflow/pull/476/

Comment (simaoribeiro) [reply src/KafkaFlow.IntegrationTests/Core/Middlewares/GzipMiddleware.cs]: Done, please review

File: src/KafkaFlow.Abstractions/IEvent.cs [hunk 3]
```diff
@@ -9,9 +9,9 @@
     public interface IEvent
     {
         /// <summary>
-        /// Subscribes to the event.
+        /// Subscribes to the async event.
```
Comment (ailtonguitar) [inline src/KafkaFlow.Abstractions/IEvent.cs:12]: Maybe we could leave the summary as before.

File: src/KafkaFlow.Abstractions/IEvent.cs [hunk 4]
```diff
@@ -9,9 +9,9 @@
...
         /// </summary>
-        /// <param name="handler">The handler to be called when the event is fired.</param>
+        /// <param name="handler">The handler to be called when the async event is fired.</param>
         /// <returns>Event subscription reference</returns>
```
Comment (ailtonguitar) [inline src/KafkaFlow.Abstractions/IEvent.cs:15]: Maybe we could leave the summary as before.

File: src/KafkaFlow.Abstractions/IEvent.cs [hunk 5]
```diff
@@ -23,9 +23,9 @@ public interface IEvent
     public interface IEvent<out TArg>
     {
         /// <summary>
-        /// Subscribes to the event.
+        /// Subscribes to the async event.
         /// </summary>
-        /// <param name="handler">The handler to be called when the event is fired.</param>
+        /// <param name="handler">The handler to be called when the async event is fired.</param>
```
Comment (ailtonguitar) [inline src/KafkaFlow.Abstractions/IEvent.cs:28]: also here

File: src/KafkaFlow.Admin.Dashboard/Properties/launchSettings.json [hunk 6]
```diff
@@ -0,0 +1,12 @@
+{
+  "profiles": {
+    "KafkaFlow.Admin.Dashboard": {
+      "commandName": "Project",
+      "launchBrowser": true,
```
Comment (ailtonguitar) [inline src/KafkaFlow.Admin.Dashboard/Properties/launchSettings.json:5]: Why this change? Can you remove that?

File: src/KafkaFlow/Event.cs [hunk 7]
```diff
@@ -2,13 +2,14 @@
...
     using System.Collections.Generic;
+    using System.Linq;
     using System.Threading.Tasks;
 
     internal class Event<TArg> : IEvent<TArg>
     {
         private readonly ILogHandler logHandler;
 
-        private readonly List<Func<TArg, Task>> handlers = new();
+        private readonly List<Func<TArg, Task>> asyncHandlers = new();
```
Comment (ailtonguitar) [inline src/KafkaFlow/Event.cs:12]:
```suggestion
        private readonly List<Func<TArg, Task>> handlers = new();
```

File: src/KafkaFlow/Event.cs [hunk 8]
```diff
@@ -17,32 +18,45 @@ public Event(ILogHandler logHandler)
 
         public IEventSubscription Subscribe(Func<TArg, Task> handler)
         {
-            if (!this.handlers.Contains(handler))
+            if (!this.asyncHandlers.Contains(handler))
```
Comment (ailtonguitar) [inline src/KafkaFlow/Event.cs:21]:
```suggestion
            if (!this.handlers.Contains(handler))
```

File: src/KafkaFlow/Event.cs [hunk 9]
```diff
@@ -17,32 +18,45 @@ public Event(ILogHandler logHandler)
...
-            return new EventSubscription(() => this.handlers.Remove(handler));
+            return new EventSubscription(() => this.asyncHandlers.Remove(handler));
         }
 
-        internal async Task FireAsync(TArg arg)
+        internal Task FireAsync(TArg arg)
         {
-            foreach (var handler in this.handlers)
+            var tasks = this.asyncHandlers
+                .Where(h => h is not null)
```
Comment (ailtonguitar) [inline src/KafkaFlow/Event.cs:32]: Is it supposed to have null handlers?

**Review by joelfoliveira — APPROVED** [review]

**Review by brmagadutra — APPROVED** [review]

File: src/KafkaFlow.Abstractions/IEvent.cs (hunk 3 above)
Comment (simaoribeiro) [reply src/KafkaFlow.Abstractions/IEvent.cs]: Done, please review

File: src/KafkaFlow.Abstractions/IEvent.cs (hunk 4 above)
Comment (simaoribeiro) [reply src/KafkaFlow.Abstractions/IEvent.cs]: Done, please review

File: src/KafkaFlow.Abstractions/IEvent.cs (hunk 5 above)
Comment (simaoribeiro) [reply src/KafkaFlow.Abstractions/IEvent.cs]: Done, please review

File: src/KafkaFlow.Admin.Dashboard/Properties/launchSettings.json (hunk 6 above)
Comment (simaoribeiro) [reply src/KafkaFlow.Admin.Dashboard/Properties/launchSettings.json]: Done, please review

File: src/KafkaFlow/Event.cs (hunk 7 above)
Comment (simaoribeiro) [reply src/KafkaFlow/Event.cs]: Done, please review

File: src/KafkaFlow/Event.cs (hunk 8 above)
Comment (simaoribeiro) [reply src/KafkaFlow/Event.cs]: Done, please review

**Review by ailtonguitar — APPROVED** [review]
